### PR TITLE
Document the advanced error formatting engine

### PR DESCRIPTION
**Pull request by tek**
Based on https://github.com/tek/splain

See https://github.com/scala/scala/pull/7785

### CODE REVIEW

**Comment by tek** [comment]
I'm gonna work out the TODO now.

@SethTisue please have a provisional look!

**Comment by tek** [comment]
need some advice on how the linking is done correctly! (see error in build)

File: _overviews/compiler-options/errors.md [hunk 1]
```diff
@@ -0,0 +1,111 @@
+# Introduction
+
+An advanced mechanism for formatting type errors and inspecting missing
+implicits has been introduced in Scala 2.15.
```
Comment (lrytz) [inline _overviews/compiler-options/errors.md:4]: 2.15 🚀

**Review by lrytz — COMMENTED** [review]
This page doesn't mention `-Vtype-diffs`

File: _overviews/compiler-options/errors.md (hunk 1 above)
Comment (dwijnand) [reply _overviews/compiler-options/errors.md:4]:
```suggestion
implicits has been introduced in Scala 2.13.6.
```

**Comment by tek** [comment]
oh no I forgot to add some files to the commit yesterday :sweat_smile:

File: _overviews/compiler-options/errors.md [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+`-Vimplicits-trunc-refined` reduces the verbosity of refined types, see
+[#truncating-refined-types].
+
+# Features
+
+The error formatting engine provides the following enhancements:
+
+# Infix Types
+
+Instead of `shapeless.::[A, HNil]`, prints `A :: HNil`.
```
[the remark this answers is not in the document]
Comment (tek) [reply _overviews/compiler-options/errors.md:34]: yes. When the type constructor is binary and its `encodedName` and `decodedName` differ.

File: _overviews/compiler-options/errors.md [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+
+<!-- special consideration for `shapeless.Record` (available at [link to analyzer plugin](https://github.com/???)): -->
+
+<!-- ![foundreq_record](/resources/img/foundreq_record.jpg) -->
+
+# Implicit Resolution Chains
+
+When an implicit is not found, only the outermost error at the invocation point
+is printed by the regular error reporter. This can be expanded with the
+compiler flag `-Xlog-implicits`, but that also shows all invalid implicits for
```
[the remark this answers is not in the document]
Comment (tek) [reply _overviews/compiler-options/errors.md:46]: This refers to the old behavious of `-Xlog-implicits`, I guess it should be rephrased now

[the remark this answers is not in the document]
Comment (tek) [reply _overviews/compiler-options/errors.md]: good now?

[the remark this answers is not in the document]
Comment (dwijnand) [reply _overviews/compiler-options/errors.md]: Yep!

File: _overviews/compiler-options/errors.md [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+  VeryLongTypeName ::::
+  VeryLongTypeName ::::
+  VeryLongTypeName
+]
+```
+
+## Truncating Refined Types
+
+A type of the shape `T { type A = X; type B = Y }` will be displayed as `T
+{...}` if the option `-Vimplicits-max-refined` is set to a value `/= 0` and
```
Comment (dwijnand) [inline _overviews/compiler-options/errors.md:102]: Your Haskell is leaking here :P  Also I reworked it so it's Int.MaxValue by default.  So 0 will actually truncate it all the way I think.  I'd go with something more general along the lines of "Setting the option -Vimplicits-max-refined can be used to define the maximum length."

File: _overviews/compiler-options/errors.md [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+
+![foundreq](/resources/img/foundreq.jpg)
+
+## Implicit Resolution Chains
+
+When an implicit is not found, only the outermost error at the invocation point is printed by the regular error
+reporter.
+Previously, the flag `-Xlog-implicits` caused the compiler to print all information about processed implicits, but the
+output was highly verbose and contained all invalid implicits for parameters that have been resolved successfully.
+The flag is now an alias for `-Vimplicits`, which prints a compact list of all involved implicit instances:
```
Comment (dwijnand) [inline _overviews/compiler-options/errors.md:49]: 👨🏼‍🎨  Very pedantically, I'd say it's renamed to -Vimplicits and the old -Xlog-implicits will continue to work as a deprecated alias.  Minor itch :)

File: _overviews/compiler-options/errors.md (hunk 4 above)
Comment (tek) [reply _overviews/compiler-options/errors.md]: :sweat_smile: rephrased it.

File: _overviews/compiler-options/errors.md (hunk 5 above)
Comment (tek) [reply _overviews/compiler-options/errors.md]: please don't hold back on the pedantry!

**Review by dwijnand — COMMENTED** [review]
Awesome.  Thanks, @tek!

All good, but only to be merged... either when we merge the PR or maybe when 2.13.6 is actually out?

**Comment by tek** [comment]
Are changes to this repo deployed to the website more frequently than compiler releases?

**Comment by dwijnand** [comment]
Yeah, I think the docs are published on merge.

**Comment by tek** [comment]
then I guess it would be confusing if it were merged now

**Comment by SethTisue** [comment]
fixup: 905dbca95bbf849a313cde4654690f5b31fde11f

**Comment by tek** [comment]
@SethTisue why did you change the link? now it points to the same page as _Compiler Options_, the splain docs aren't reachable

**Comment by SethTisue** [comment]
@tek I see now that my fix was wrong, but the problem I was attempting to address is real — https://docs.scala-lang.org/overviews/compiler-options/errors.html is 404. and when I generate the site locally, no `errors.html` file is generated

The problem was that `errors.md` lacked the proper front matter. Fixed: 5f734205a00ae937f335c59e5fed26d4d6b69cf9

Thanks for catching my mistake!

**Comment by tek** [comment]
wonderful, thanks a lot!

**Pull request merged**